doc/md/tutorial-todo-gql-mutation-input.md

```diff
@@ -0,0 +1,422 @@
+generator using Go templates and generate [input type](https://graphql.org/graphql-js/mutations-and-input-types/)
+objects for our GraphQL mutations that can be applied directly on Ent mutations.
+
+#### Clone the code (optional)
+
+The code for this tutorial is available under [github.com/a8m/ent-graphql-example](https://github.com/a8m/ent-graphql-example),
+and tagged (using Git) in each step. If you want to skip the basic setup and start with the initial version of the GraphQL
+server, you can clone the repository and run the program as follows:
+
+```console
+git clone git@github.com:a8m/ent-graphql-example.git
+cd ent-graphql-example 
+go run ./cmd/todo/
+```
+
+## Go Templates
+
+The Ent framework accepts external templates that can extend or override the default generated functionality of its 
+code generator. In the template below, we generate 2 input types (`CreateTodoInput` and `UpdateTodoInput`) for the
+GraphQL mutations, and add additional methods on the different builders to accept these objects as an input type.
+
+
+```gotemplate
+{{ range $n := $.Nodes }}
+    {{ $input := print "Create" $n.Name "Input" }}
+    // {{ $input }} represents a mutation input for creating {{ plural $n.Name | lower }}.
+    type {{ $input }} struct {
+        {{- range $f := $n.Fields }}
+            {{ $f.StructField }} {{ if and (or $f.Optional $f.Default) (not $f.Type.RType.IsPtr) }}*{{ end }}{{ $f.Type }}
+        {{- end }}
+        {{- range $e := $n.Edges }}
+            {{- if $e.Unique }}
+                {{ $e.StructField }} {{ if $e.Optional }}*{{ end }}{{ $e.Type.ID.Type }}
+            {{- else }}
+                {{ $e.StructField }} []{{ $e.Type.ID.Type }}
+            {{- end }}
+        {{- end }}
+    }
+
+    {{/* Additional methods go here.  */}}
+
+    {{ $input = print "Update" $n.Name "Input" }}
+    // {{ $input }} represents a mutation input for updating {{ plural $n.Name | lower }}.
+    type {{ $input }} struct {
+        {{- range $f := $n.MutableFields }}
+            {{ $f.StructField }} {{ if not $f.Type.RType.IsPtr }}*{{ end }}{{ $f.Type }}
+            {{- if $f.Optional }}
+                {{ "Clear" $f.StructField }} bool
+            {{- end }}
+        {{- end }}
+        {{- range $e := $n.Edges }}
+            {{- if $e.Unique }}
+                {{ $e.StructField }} *{{ $e.Type.ID.Type }}
+                {{ $e.MutationClear }} bool
+            {{- else }}
+                {{ $e.MutationAdd }} []{{ $e.Type.ID.Type }}
+                {{ $e.MutationRemove }} []{{ $e.Type.ID.Type }}
+            {{- end }}
+        {{- end }}
+    }
+
+    {{/* Additional methods go here.  */}}
+{{ end }}
+```
+
+The full version of this template exists in the [repository](https://github.com/a8m/ent-graphql-example/blob/master/ent/template/mutation_input.tmpl).
+
+Now, we tell the Ent code generator to execute this template by passing it as an argument in the `ent/entc.go` file:
```

Comment:
I think it will be useful to add an "info box" explaining a bit about the different "magic" (ie available in the template FuncMap) functions that are available. Give an example or two and link to where I can find the whole thing.